lib/Log/Any/Adapter/DERIV.pm

```diff
@@ -331,6 +339,63 @@ sub _in_container{
     return -r '/.dockerenv';
 }
 
+=head2 _linux_flock
+
+Param: lock type. It can be F_WRLCK or F_UNLCK
+
+return: A FLOCK structure
+
+=cut
+
+# The following code is from https://docstore.mik.ua/orelly/perl4/cook/ch07_26.htm
+sub _linux_flock {
+    my ($type) = @_;
+    my $FLOCK_STRUCT = "s s l l i";
+    return pack($FLOCK_STRUCT, $type, SEEK_SET, 0, 0, 0);
+}
+
+=head2 _lock
+
+Lock a file handler with fcntl.
+
+Param: fh - File handle
+
+Return: true or false
+
+=cut
+
+# We don't use `flock` function directly here
+# In some cases the program will do fork after the log file opened.
+# In such case every subprocess can get lock of the log file at the same time.
+# Using fcntl to lock a file can avoid this problem
+sub _lock{
+    my ($fh) = @_;
+    my $lock = _linux_flock(F_WRLCK);
+    my $result = fcntl($fh, F_SETLKW, $lock);
+    return $result if $result;
+    print STDERR "F_SETLKW @_: $!\n";
+    return undef;
+}
+
+=head2 _unlock
+
+Unlock a file handler locked by fcntl
+
+Param: fh - File handle
+Return: true or false
+
+=cut
+
+sub _unlock{
+    my ($fh) = @_;
+    my $lock = _linux_flock(F_UNLCK);
+    my $result = fcntl($fh, F_SETLKW, $lock);
+    return $result if $result;
+    print STDERR "F_SETLKW @_: $!\n";
+    return undef;
+}
+
```

Comment:
_lock and _unlock subroutine varies only on F_WRLCK and F_UNLCK respectively. could create  something like below (move the common code to another sub called _flock)

```
sub _flock{
    my ($fh, $type) = @_;
    my $lock = _linux_flock($type);
    my $result = fcntl($fh, F_SETLKW, $lock);
    return $result if $result;
    print STDERR "F_SETLKW @_: $!\n";
    return undef;
}

sub _lock{
    my ($fh) = @_;
    return _flock($fh, F_WRLCK);
}
sub _unlock{
    my ($fh) = @_;
    return _flock(@fh, F_UNLCK);
}
```